fix(aio): intercept all clicks on anchors

petebacondarwin opened the pull request:
Previously we had the `LinkDirective` which intercepted clicks on anchors
outside the doc viewer. Now we intercept "all" link clicks within the app.

Builds on top of #15040

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

Clicks on links within the doc-viewer are not being intercepted, which leads to unwanted page reloads.

**What is the new behavior?**

Clicks are now being intercepted.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

Links that start with a `#` are still not dealt with properly. They could be intercepted and mapped in the same way as the other links but that would not deal with the actual href being broken from the view point of being able to cut and paste the link to share (or to open the link in a new browser window). This needs to be fixed in the dgeni pipeline.

mary-poppins commented:
The angular.io preview for 4b61c11 is available [here][1].

[1]: https://pr15083-4b61c112efd02ad1b2382e68cb42d113d3dd5517.ngbuilds.io/

petebacondarwin commented on aio/src/app/app.component.ts:
```diff
@@ -20,7 +21,10 @@ export class AppComponent implements OnInit {
+  constructor(documentService: DocumentService,
+              navigationService: NavigationService,
+              private searchService: SearchService,
+              private locationService: LocationService) {
```
Took a leaf out of @wardbell's book here :-)

petebacondarwin commented on aio/src/app/app.component.ts:
```diff
@@ -41,4 +45,12 @@ export class AppComponent implements OnInit {
+    if (eventTarget instanceof HTMLAnchorElement) {
+      return this.locationService.handleAnchorClick(eventTarget, button, ctrlKey, metaKey);
+    }
+    return true;
```
The click catching functionality is separated into here and the `LocationService` for three reasons.

1) It makes testing easier, since I can mock out `handleAnchorClick` to always return `false` which prevents Karma from getting confused by real navigations due to clicking on links.
2) The handling of anchor clicks requires intimate knowledge of the `LocationService` which I didn't want to expose here
3) I am going to add in a click handler for closing the search results in here, since we need it to be done somewhere that has knowledge of both the results view and the search box. These are separate components in different parts of the app, and we only want to close the results if you click outside of both components.

gkalpak replied:
I would still expect the functionlity to be in a directive (attached to `<a>` elements). Your points above don't sound very convincing to me (😛  - it's porbably me, not you), but [this line] does:

> in this app we have a special situation where the `DocViewerComponent` is displaying semi-static content that cannot contain directives.

[1]: https://github.com/angular/angular/pull/15083/files#diff-d7053af3a910e9defc92631371cb06efR72

petebacondarwin replied:
The comments in here are why I have split the functionality out of the `AppComponent`, not why it is in the `AppComponent` in the first place.

gkalpak replied:
See? I told you it was me 😛

petebacondarwin commented on aio/src/app/app.module.ts:
```diff
@@ -27,7 +27,6 @@ import { SearchService } from 'app/search/search.service';
 import { TopMenuComponent } from 'app/layout/top-menu/top-menu.component';
 import { NavMenuComponent } from 'app/layout/nav-menu/nav-menu.component';
 import { NavItemComponent } from 'app/layout/nav-item/nav-item.component';
-import { LinkDirective } from 'app/shared/link.directive';
```
The new functionality replaces the `LinkDirective` which never worked inside the doc views anyway.

petebacondarwin commented on aio/src/app/shared/location.service.ts:
```diff
@@ -59,4 +61,31 @@ export class LocationService {
+    const { host, pathname, search, hash } = anchor;
+    const relativeUrl = pathname + search + hash;
+    this.urlParser.href = relativeUrl;
+    if (anchor.href !== this.urlParser.href) {
```
I am hoping that this technique works for all scenarios. The idea is that we extract the relative bits and attach it to another link. If all the absolute bits are the same then we should be OK.

There is a chance that changes to the base href might break this but I haven't found it yet.

gkalpak commented on aio/src/app/shared/location.service.ts:
```diff
@@ -59,4 +61,31 @@ export class LocationService {
+    }
+
+    // check for external link
+    const { host, pathname, search, hash } = anchor;
```
`host` doesn't seem to be used.

gkalpak reviewed (commented):
I wonder why not implement it as a directive attached to `<a>` elements. Having it in `AppComponent` feels a bit _spooky-action-at-a-distance-y_ (unless there is good reason :smiley:).

Other than that LGTM.

petebacondarwin commented:
I'll add some documentation to explain why the `LinkDirective` didn't cut it.

mary-poppins commented:
The angular.io preview for 8458a12 is available [here][1].

[1]: https://pr15083-8458a1227e7e6f66fe4c2d58243281ba4fab614c.ngbuilds.io/

gkalpak reviewed (approved):
LGTM

petebacondarwin committed:
fix(aio): ignore private exports

Closes #14992

petebacondarwin committed:
refactor(aio): move window:resize handler into component

mary-poppins commented:
The angular.io preview for 48b4f90 is available [here][1].

[1]: https://pr15083-48b4f909473a7fb1dcbc4b0a7677e56ed25b985b.ngbuilds.io/

petebacondarwin committed:
fix(aio): intercept all clicks on anchors

Previously we had the `LinkDirective` which intercepted clicks on anchors
outside the doc viewer. Now we intercept "all" link clicks within the app.

mary-poppins commented:
The angular.io preview for 994d668 is available [here][1].

[1]: https://pr15083-994d6687710ba3dd8eedf03ce7ddbb335fd501a6.ngbuilds.io/
